doc/monads/functors.lean

```diff
@@ -0,0 +1,215 @@
+/-!
+# Functor
+
+A `Functor` is any type that can act as a generic container that allows you to transform the
+underlying values inside the container using a function, so that the values are all updated, but the
+structure of the container is the same. This is called "mapping".
+
+A List is the most basic example of a `Functor`.
+
+A list contains zero or more elements of the same, underlying type.  When you `map` a function over
+a list, you create a new list with the same number of elements, where each has been transformed by
+the function:
+-/
+#eval [1,2,3].map (λ x => toString x) -- ["1", "2", "3"]
+
+/-!
+Here we converted a list of natural numbers (Nat) to a list of strings where the lambda function
+here used `toString` to do the transformation of each element. Notice that when you apply `map`
+the "structure" of the object remains the same, in this case the resulting List is the same size.
+
+List has a specialized version of `map` defined as follows:
+-/
+def map (f : α → β) : List α → List β
+  | []    => []
+  | a::as => f a :: map f as
+
+/-!
+This is a very generic `map` function that can take any function that converts `(α → β)` and use it
+to convert `List α → List β`. Notice the function call `f a` above, this application of `f`
+is producing the converted items for the new list.
+
+Lean also defines custom infix operator `<$>` for map which simply allows you to write this:
+-/
+#eval (λ x => toString x) <$> [1, 2, 3] -- ["1", "2", "3"]
+/-!
+
+In this `Functor`, `map` is called a "higher order function" because it takes a function as input
+`(α → β)` and returns another function as output `F α → F β`.
+
+Note that Lean is very powerful and lets you define your own syntax, so `<$>` is nothing special.
+You can define your own infix operator like this:
+
+-/
+infixr:100 " doodle " => Functor.map
+
+#eval (· * 5) doodle [1, 2, 3]  -- [5, 10, 15]
+/-!
+
+Let's look at some more examples:
+
+-/
+-- List String → List Nat
+#eval ["elephant", "tiger", "giraffe"].map (fun s => s.length)
```

Comment:
We can say that we are using `List.map`, and remind the reader that we are using the dot notation. Not sure what the best approach is.

Reply:
That's fine the earlier snippet was not trying to redefine List.map it was just showing the reader what List.map is.

Reply:
It does not look fine to me. We should at least say it. Otherwise, the reader may assume that after you define a function called `map`, they can write `[1,2,3,4,5].map ...`

Reply:
I'm still not sure I agree, notice I use the list `.map` function BEFORE I show the map implementation on this line: 
```lean
#eval [1,2,3].map (λ x => toString x) -- ["1", "2", "3"]
```
And it clearly introduces the code as "List has a specialized version of `map` defined as follows:" this is pretty clearly saying this is what map does, and it is not saying something like "Let's define a new map function like this:" then sure, we'd have to show the full Functor redefinition, but that would be jumping the gun because this is the chapter on Functors which we haven't introduced yet...

So I move to the "dot" notation because of your comments on List and Array docs should use dot notation.  Would it be better if I moved these early snippets back to using `List.map` to make it clear which version of map we are using here?

Reply:
Yes, there are four different `map` functions in this section: `List.map`, `Option.map`, `Functor.map`, and `map` (the version of `List.map` you have defined here). The notation for `<$>` is for `Functor.map`.

Reply:
Ok, I added `List.map` and moved the reference to `<$>` down and introduced it like this:

> Lean also defines custom infix operator `<$>` for `Functor.map` which allows you to write this:

See https://github.com/lovettchris/lean4/commit/2ef1e40fc5eae9a452a5d73af6cd065c9e352a69